File: eedc/frontend/src/api/haStatistics.ts
/**
 * HA Statistics API Client
 *
 * Ermöglicht den Zugriff auf Home Assistant Langzeitstatistiken
 * für historischen Datenimport.
 */

import { api } from './client'

// =============================================================================
// Types
// =============================================================================

export interface HAStatisticsStatus {
  verfuegbar: boolean
  pfad?: string
  fehler?: string
}

export interface FeldWert {
  feld: string
  label: string
  wert: number | null
  einheit: string
  sensor_id?: string
}

// Backend Response Format für MappedMonatswert
interface BackendMappedFeld {
  feld: string
  feld_label: string
  sensor_id: string
  start_wert: number
  end_wert: number
  differenz: number
  einheit: string
}

// Backend Response Format für InvestitionMitFelder
interface BackendInvestitionMitFelder {
  investition_id: number
  bezeichnung: string
  typ: string
  felder: BackendMappedFeld[]
}

// Backend Response Format für AnlagenMonatswertResponse
interface BackendMonatswerte {
  anlage_id: number
  anlage_name: string
  jahr: number
  monat: number
  monat_name: string
  basis: BackendMappedFeld[]
  investitionen: BackendInvestitionMitFelder[]  // Liste von Investitionen mit Feldern
}

export interface InvestitionWerte {
  investition_id: number
  bezeichnung: string
  typ: string
  felder: FeldWert[]
}

export interface Monatswerte {
  jahr: number
  monat: number
  monat_name: string
  basis: FeldWert[]
  investitionen: InvestitionWerte[]
}

/**
 * Konvertiert Backend-Format zu Frontend-Format
 */
function convertBackendToMonatswerte(backend: BackendMonatswerte): Monatswerte {
  return {
    jahr: backend.jahr,
    monat: backend.monat,
    monat_name: backend.monat_name,
    basis: backend.basis.map(b => ({
      feld: b.feld,
      label: b.feld_label,
      wert: b.differenz,
      einheit: b.einheit,
      sensor_id: b.sensor_id
    })),
    investitionen: backend.investitionen.map(inv => ({
      investition_id: inv.investition_id,
      bezeichnung: inv.bezeichnung,
      typ: inv.typ,
      felder: inv.felder.map(f => ({
        feld: f.feld,
        label: f.feld_label,
        wert: f.differenz,
        einheit: f.einheit,
        sensor_id: f.sensor_id
      }))
    }))
  }
}

export interface VerfuegbarerMonat {
  jahr: number
  monat: number
  monat_name: string
  hat_daten?: boolean  // Optional, nicht immer vom Backend geliefert
}

export interface VerfuegbareMonateResponse {
  anlage_id: number
  anlage_name: string
  erstes_datum: string
  letztes_datum: string
  anzahl_monate: number
  monate: VerfuegbarerMonat[]
}

export interface AlleMonatswerte {
  anlage_id: number
  anlage_name: string
  monate: Monatswerte[]
  sensor_count: number
  zeitraum?: {
    von: string
    bis: string
  }
}

export interface MonatsanfangWerte {
  anlage_id: number
  anlage_name: string
  jahr: number
  monat: number
  startwerte: Record<string, number>
}

export type ImportAktion = 'importieren' | 'ueberspringen' | 'ueberschreiben' | 'konflikt'

export interface InvestitionImportStatus {
  investition_id: number
  bezeichnung: string
  typ: string
  ha_werte: Record<string, number | null>
  vorhandene_werte: Record<string, number | null>
  hat_abweichung: boolean
}

export interface MonatImportStatus {
  jahr: number
  monat: number
  monat_name: string
  aktion: ImportAktion
  grund: string
  ha_werte: Record<string, number | null>
  vorhandene_werte?: Record<string, number | null>
  investitionen?: InvestitionImportStatus[]
}

export interface ImportVorschau {
  anlage_id: number
  anlage_name: string
  anzahl_monate: number
  anzahl_importieren: number
  anzahl_konflikte: number
  anzahl_ueberspringen: number
  monate: MonatImportStatus[]
}

export interface MonatFeldAuswahl {
  jahr: number
  monat: number
  // Basis-Felder: Liste der Feld-Namen die importiert werden sollen
  // null/undefined = alle, [] = keine
  basis_felder?: string[] | null
  // Investitions-Felder: Dict von inv_id -> Liste der Feld-Namen
  // null/undefined = alle, {} = keine
  investition_felder?: Record<string, string[]> | null
}

export interface ImportRequest {
  monate: MonatFeldAuswahl[]
  ueberschreiben: boolean
}

export interface ImportResult {
  erfolg: boolean
  importiert: number
  uebersprungen: number
  ueberschrieben: number
  fehler: string[]
}

// =============================================================================
// API Client
// =============================================================================

export const haStatisticsApi = {
  /**
   * Prüft ob HA-Datenbank verfügbar ist
   */
  async getStatus(): Promise<HAStatisticsStatus> {
    return api.get<HAStatisticsStatus>('/ha-statistics/status')
  },

  /**
   * Monatswerte für einen einzelnen Monat abrufen
   */
  async getMonatswerte(
    anlageId: number,
    jahr: number,
    monat: number
  ): Promise<Monatswerte> {
    const backend = await api.get<BackendMonatswerte>(`/ha-statistics/monatswerte/${anlageId}/${jahr}/${monat}`)
    return convertBackendToMonatswerte(backend)
  },

  /**
   * Alle verfügbaren Monate mit Daten abrufen
   */
  async getVerfuegbareMonate(anlageId: number): Promise<VerfuegbarerMonat[]> {
    const response = await api.get<VerfuegbareMonateResponse>(`/ha-statistics/verfuegbare-monate/${anlageId}`)
    // Backend gibt Wrapper-Objekt zurück, wir extrahieren nur die monate-Liste
    return response.monate.map(m => ({
      ...m,
      hat_daten: true  // Alle zurückgegebenen Monate haben Daten
    }))
  },

  /**
   * Alle Monatswerte (Bulk-Abfrage)
   */
  async getAlleMonatswerte(anlageId: number): Promise<AlleMonatswerte> {
    return api.get<AlleMonatswerte>(`/ha-statistics/alle-monatswerte/${anlageId}`)
  },

  /**
   * Zählerstände am Monatsanfang (für MQTT Startwerte)
   */
  async getMonatsanfangWerte(
    anlageId: number,
    jahr: number,
    monat: number
  ): Promise<MonatsanfangWerte> {
    return api.get<MonatsanfangWerte>(`/ha-statistics/monatsanfang/${anlageId}/${jahr}/${monat}`)
  },

  /**
   * Import-Vorschau mit Konflikt-Erkennung
   */
  async getImportVorschau(anlageId: number): Promise<ImportVorschau> {
    return api.get<ImportVorschau>(`/ha-statistics/import-vorschau/${anlageId}`)
  },

  /**
   * Import durchführen mit Überschreib-Schutz
   */
  async importieren(anlageId: number, request: ImportRequest): Promise<ImportResult> {
    return api.post<ImportResult>(`/ha-statistics/import/${anlageId}`, request)
  },
}

File: eedc/backend/api/routes/ha_statistics.py
"""
HA Statistics API Routes

Endpoints für den Zugriff auf Home Assistant Langzeitstatistiken.

Ermöglicht:
- Monatswerte für einen bestimmten Monat abrufen
- Alle verfügbaren Monate ermitteln
- Bulk-Import aller historischen Monatswerte
- MQTT-Startwerte initialisieren
"""

from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel

from backend.api.deps import get_db
from backend.models.anlage import Anlage
from backend.services.ha_statistics_service import (
    get_ha_statistics_service,
    MonatswertResponse,
    AlleMonateResponse,
    SensorMonatswert,
)

router = APIRouter()


# =============================================================================
# Response Models
# =============================================================================

class MappedMonatswert(BaseModel):
    """Monatswert mit EEDC-Feld-Zuordnung."""
    feld: str
    feld_label: str
    sensor_id: str
    start_wert: float
    end_wert: float
    differenz: float
    einheit: str = "kWh"


class AnlagenMonatswertResponse(BaseModel):
    """Monatswerte für eine Anlage mit Feld-Zuordnung."""
    anlage_id: int
    anlage_name: str
    jahr: int
    monat: int
    monat_name: str
    basis: list[MappedMonatswert]
    investitionen: dict[str, list[MappedMonatswert]]  # investition_id -> felder


class VerfuegbareMonate(BaseModel):
    """Verfügbare Monate für eine Anlage."""
    anlage_id: int
    anlage_name: str
    erstes_datum: date
    letztes_datum: date
    anzahl_monate: int
    monate: list[dict]  # [{"jahr": 2024, "monat": 10, "monat_name": "Oktober"}, ...]


class StatusResponse(BaseModel):
    """Status der HA-Statistics-Integration."""
    verfuegbar: bool
    db_pfad: Optional[str]
    hinweis: str


# =============================================================================
# Feld-Labels für bessere Lesbarkeit
# =============================================================================

FELD_LABELS = {
    # Basis
    "einspeisung": "Einspeisung",
    "netzbezug": "Netzbezug",
    "pv_gesamt": "PV Erzeugung Gesamt",
    # PV-Module
    "pv_erzeugung_kwh": "PV Erzeugung",
    # Speicher
    "ladung_kwh": "Ladung",
    "entladung_kwh": "Entladung",
    "ladung_netz_kwh": "Ladung aus Netz",
    # E-Auto
    "ladung_pv_kwh": "Ladung PV",
    "km_gefahren": "Kilometer",
    "v2h_entladung_kwh": "V2H Entladung",
    # Wärmepumpe
    "stromverbrauch_kwh": "Stromverbrauch",
    "heizenergie_kwh": "Heizenergie",
    "warmwasser_kwh": "Warmwasser",
}


# =============================================================================
# Helper Functions
# =============================================================================

async def get_anlage_with_mapping(db: AsyncSession, anlage_id: int) -> Anlage:
    """Holt Anlage und prüft ob sensor_mapping vorhanden."""
    result = await db.execute(select(Anlage).where(Anlage.id == anlage_id))
    anlage = result.scalar_one_or_none()

    if not anlage:
        raise HTTPException(status_code=404, detail=f"Anlage {anlage_id} nicht gefunden")

    if not anlage.sensor_mapping:
        raise HTTPException(
            status_code=400,
            detail="Keine Sensor-Zuordnung konfiguriert. Bitte zuerst Sensoren zuordnen."
        )

    return anlage


def extract_sensor_ids_from_mapping(sensor_mapping: dict) -> list[str]:
    """Extrahiert alle sensor_ids aus dem Mapping."""
    sensor_ids = []

    # Basis-Sensoren
    basis = sensor_mapping.get("basis", {})
    for feld, config in basis.items():
        if config and config.get("strategie") == "sensor" and config.get("sensor_id"):
            sensor_ids.append(config["sensor_id"])

    # Investitions-Sensoren
    investitionen = sensor_mapping.get("investitionen", {})
    for inv_id, inv_config in investitionen.items():
        felder = inv_config.get("felder", {})
        for feld, config in felder.items():
            if config and config.get("strategie") == "sensor" and config.get("sensor_id"):
                sensor_ids.append(config["sensor_id"])

    return sensor_ids


def map_sensor_values_to_fields(
    sensor_mapping: dict,
    sensoren: list[SensorMonatswert]
) -> tuple[list[MappedMonatswert], dict[str, list[MappedMonatswert]]]:
    """
    Ordnet Sensorwerte den EEDC-Feldern zu.

    Returns:
        Tuple von (basis_felder, investitions_felder)
    """
    # Sensor-Werte als Dict für schnellen Zugriff
    sensor_values = {s.sensor_id: s for s in sensoren}

    basis_felder: list[MappedMonatswert] = []
    inv_felder: dict[str, list[MappedMonatswert]] = {}

    # Basis-Felder
    basis = sensor_mapping.get("basis", {})
    for feld, config in basis.items():
        if config and config.get("strategie") == "sensor":
            sensor_id = config.get("sensor_id")
            if sensor_id and sensor_id in sensor_values:
                sv = sensor_values[sensor_id]
                basis_felder.append(MappedMonatswert(
                    feld=feld,
                    feld_label=FELD_LABELS.get(feld, feld),
                    sensor_id=sensor_id,
                    start_wert=sv.start_wert,
                    end_wert=sv.end_wert,
                    differenz=sv.differenz,
                    einheit=sv.einheit
                ))

    # Investitions-Felder
    investitionen = sensor_mapping.get("investitionen", {})
    for inv_id, inv_config in investitionen.items():
        felder = inv_config.get("felder", {})
        inv_felder[inv_id] = []

        for feld, config in felder.items():
            if config and config.get("strategie") == "sensor":
                sensor_id = config.get("sensor_id")
                if sensor_id and sensor_id in sensor_values:
                    sv = sensor_values[sensor_id]
                    inv_felder[inv_id].append(MappedMonatswert(
                        feld=feld,
                        feld_label=FELD_LABELS.get(feld, feld),
                        sensor_id=sensor_id,
                        start_wert=sv.start_wert,
                        end_wert=sv.end_wert,
                        differenz=sv.differenz,
                        einheit=sv.einheit
                    ))

    return basis_felder, inv_felder


# =============================================================================
# Endpoints
# =============================================================================

@router.get("/status", response_model=StatusResponse)
async def get_ha_statistics_status():
    """
    Prüft ob HA-Statistik-Abfrage verfügbar ist.

    Returns:
        Status der Integration
    """
    service = get_ha_statistics_service()

    if service.is_available:
        return StatusResponse(
            verfuegbar=True,
            db_pfad=str(service.db_path),
            hinweis="HA-Datenbank verfügbar. Statistik-Abfragen möglich."
        )
    else:
        return StatusResponse(
            verfuegbar=False,
            db_pfad=None,
            hinweis="HA-Datenbank nicht gefunden. Nur im HA-Addon verfügbar."
        )


@router.get("/monatswerte/{anlage_id}/{jahr}/{monat}", response_model=AnlagenMonatswertResponse)
async def get_monatswerte(
    anlage_id: int,
    jahr: int,
    monat: int,
    db: AsyncSession = Depends(get_db)
):
    """
    Holt Monatswerte für einen bestimmten Monat aus HA-Statistiken.

    Args:
        anlage_id: ID der Anlage
        jahr: Jahr (z.B. 2024)
        monat: Monat (1-12)

    Returns:
        Monatswerte für alle gemappten Sensoren
    """
    if monat < 1 or monat > 12:
        raise HTTPException(status_code=400, detail="Monat muss zwischen 1 und 12 liegen")

    service = get_ha_statistics_service()
    if not service.is_available:
        raise HTTPException(
            status_code=503,
            detail="HA-Datenbank nicht verfügbar. Diese Funktion ist nur im HA-Addon nutzbar."
        )

    anlage = await get_anlage_with_mapping(db, anlage_id)
    sensor_ids = extract_sensor_ids_from_mapping(anlage.sensor_mapping)

    if not sensor_ids:
        raise HTTPException(
            status_code=400,
            detail="Keine Sensor-Zuordnungen mit Strategie 'sensor' gefunden."
        )

    # Werte aus HA-DB holen
    try:
        response = service.get_monatswerte(sensor_ids, jahr, monat)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Fehler bei DB-Abfrage: {e}")

    # Auf EEDC-Felder mappen
    basis_felder, inv_felder = map_sensor_values_to_fields(
        anlage.sensor_mapping,
        response.sensoren
    )

    return AnlagenMonatswertResponse(
        anlage_id=anlage.id,
        anlage_name=anlage.anlagenname,
        jahr=jahr,
        monat=monat,
        monat_name=response.monat_name,
        basis=basis_felder,
        investitionen=inv_felder
    )


@router.get("/verfuegbare-monate/{anlage_id}", response_model=VerfuegbareMonate)
async def get_verfuegbare_monate(
    anlage_id: int,
    db: AsyncSession = Depends(get_db)
):
    """
    Ermittelt alle Monate mit verfügbaren HA-Statistik-Daten.

    Args:
        anlage_id: ID der Anlage

    Returns:
        Liste aller Monate mit Daten
    """
    service = get_ha_statistics_service()
    if not service.is_available:
        raise HTTPException(
            status_code=503,
            detail="HA-Datenbank nicht verfügbar. Diese Funktion ist nur im HA-Addon nutzbar."
        )

    anlage = await get_anlage_with_mapping(db, anlage_id)
    sensor_ids = extract_sensor_ids_from_mapping(anlage.sensor_mapping)

    if not sensor_ids:
        raise HTTPException(
            status_code=400,
            detail="Keine Sensor-Zuordnungen mit Strategie 'sensor' gefunden."
        )

    try:
        response = service.get_verfuegbare_monate(sensor_ids)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Fehler bei DB-Abfrage: {e}")

    return VerfuegbareMonate(
        anlage_id=anlage.id,
        anlage_name=anlage.anlagenname,
        erstes_datum=response.erstes_datum,
        letztes_datum=response.letztes_datum,
        anzahl_monate=response.anzahl_monate,
        monate=[
            {
                "jahr": m.jahr,
                "monat": m.monat,
                "monat_name": m.monat_name
            }
            for m in response.monate
        ]
    )


@router.get("/alle-monatswerte/{anlage_id}", response_model=list[AnlagenMonatswertResponse])
async def get_alle_monatswerte(
    anlage_id: int,
    ab_jahr: Optional[int] = Query(None, description="Nur Monate ab diesem Jahr"),
    ab_monat: Optional[int] = Query(None, description="Nur Monate ab diesem Monat (mit ab_jahr)"),
    db: AsyncSession = Depends(get_db)
):
    """
    Holt alle verfügbaren Monatswerte aus HA-Statistiken.

    Ideal für:
    - Initialbefüllung bei neuer Installation
    - Nachträgliche Korrektur fehlender Monate
    - Bulk-Import historischer Daten

    Args:
        anlage_id: ID der Anlage
        ab_jahr: Optional - nur Monate ab diesem Jahr
        ab_monat: Optional - nur Monate ab diesem Monat (erfordert ab_jahr)

    Returns:
        Liste aller Monatswerte
    """
    service = get_ha_statistics_service()
    if not service.is_available:
        raise HTTPException(
            status_code=503,
            detail="HA-Datenbank nicht verfügbar. Diese Funktion ist nur im HA-Addon nutzbar."
        )

    anlage = await get_anlage_with_mapping(db, anlage_id)
    sensor_ids = extract_sensor_ids_from_mapping(anlage.sensor_mapping)

    if not sensor_ids:
        raise HTTPException(
            status_code=400,
            detail="Keine Sensor-Zuordnungen mit Strategie 'sensor' gefunden."
        )

    # Ab-Datum berechnen
    ab_datum = None
    if ab_jahr:
        ab_datum = date(ab_jahr, ab_monat or 1, 1)

    try:
        raw_responses = service.get_alle_monatswerte(sensor_ids, ab_datum)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Fehler bei DB-Abfrage: {e}")

    # Auf EEDC-Felder mappen
    ergebnisse = []
    for raw in raw_responses:
        basis_felder, inv_felder = map_sensor_values_to_fields(
            anlage.sensor_mapping,
            raw.sensoren
        )

        ergebnisse.append(AnlagenMonatswertResponse(
            anlage_id=anlage.id,
            anlage_name=anlage.anlagenname,
            jahr=raw.jahr,
            monat=raw.monat,
            monat_name=raw.monat_name,
            basis=basis_felder,
            investitionen=inv_felder
        ))

    return ergebnisse


@router.get("/monatsanfang/{anlage_id}/{jahr}/{monat}")
async def get_monatsanfang_werte(
    anlage_id: int,
    jahr: int,
    monat: int,
    db: AsyncSession = Depends(get_db)
):
    """
    Holt die Zählerstände am Monatsanfang.

    Nützlich für die Initialisierung der MQTT-Startwerte.

    Args:
        anlage_id: ID der Anlage
        jahr: Jahr
        monat: Monat (1-12)

    Returns:
        Dict mit Zählerständen pro Sensor
    """
    if monat < 1 or monat > 12:
        raise HTTPException(status_code=400, detail="Monat muss zwischen 1 und 12 liegen")

    service = get_ha_statistics_service()
    if not service.is_available:
        raise HTTPException(
            status_code=503,
            detail="HA-Datenbank nicht verfügbar. Diese Funktion ist nur im HA-Addon nutzbar."
        )

    anlage = await get_anlage_with_mapping(db, anlage_id)
    sensor_ids = extract_sensor_ids_from_mapping(anlage.sensor_mapping)

    startwerte = {}
    for sensor_id in sensor_ids:
        wert = service.get_monatsanfang_wert(sensor_id, jahr, monat)
        if wert is not None:
            startwerte[sensor_id] = wert

    return {
        "anlage_id": anlage.id,
        "anlage_name": anlage.anlagenname,
        "jahr": jahr,
        "monat": monat,
        "startwerte": startwerte
    }
